### PR TITLE
Sql editor compatibility + improved pattern geometry editing

### PR DESCRIPTION
**Pull request by landonreed**
This PR updates the UI to be compatible with the new SQL-based editor.  A list of changes includes:
- Converts a number of actions to use `redux-actions`
- GraphQL fetch for for all editor entities
- Change get/update/delete single entity actions to be generic (shared action logic that handles changing the endpoint and mapping for each entity type, e.g., `PUT /api/editor/${entityType}`
- Updates the pattern geometry editing logic:
  - Server now sends list of shape points
  - UI now handles projecting stops onto the pattern shape
  - Control points (anchor points) are persisted in the back end
  - Single, inter-stop (or inter-control point) segments are independently editable/visible to improve editing for loop routes or routes that double back on themselves.

### CODE REVIEW

File: lib/editor/actions/map/stopStrategies.js [hunk 1]
```diff
@@ -1,16 +1,18 @@
-import along from 'turf-along'
+import along from '@turf/along'
 import ll from '@conveyal/lonlat'
 import lineDistance from 'turf-line-distance'
 import lineSlice from 'turf-line-slice'
```
Comment (evansiroky) [inline lib/editor/actions/map/stopStrategies.js]: We should try to use the latest turf packages

File: lib/common/actions/index.js [hunk 2]
```diff
@@ -1,37 +1,44 @@
 import fetch from 'isomorphic-fetch'
+import {GTFS_GRAPHQL_PREFIX} from '../../common/constants'
 import { setErrorMessage } from '../../manager/actions/status'
 
-export function secureFetch (url, method = 'get', payload, raw) {
+export function secureFetch (url, method = 'get', payload, raw = false, isJSON = true) {
```
Comment (evansiroky) [inline lib/common/actions/index.js]: I'm seeing a lot of `res.json().then(` in other areas.  Perhaps we can add something that assumes a response is json or assume that if `raw` is false, then json should be returned.

File: lib/editor/actions/tripPattern.js [hunk 3]
```diff
@@ -1,196 +1,69 @@
...
-  }
-}
+const requestingTripPatterns = createAction('REQUESTING_TRIP_PATTERNS')
+const receiveTripPatterns = createAction('RECEIVE_TRIP_PATTERNS')
+export const undoActiveTripPatternEdits = createAction('UNDO_TRIP_PATTERN_EDITS')
+export const setActiveStop = createAction('SET_ACTIVE_PATTERN_STOP')
+export const togglingPatternEditing = createAction('TOGGLE_PATTERN_EDITING')
+export const setActivePatternSegment = createAction('SET_ACTIVE_PATTERN_SEGMENT')
+const savedTripPattern = createAction('SAVED_TRIP_PATTERN')
+export const resnapStops = createAction('RESNAP_STOPS')
```
Comment (evansiroky) [inline lib/editor/actions/tripPattern.js]: Really stoked to see all this work to convert to redux-actions.  👍

File: lib/editor/util/gtfs.js [hunk 4]
```diff
@@ -1,37 +1,88 @@
...
+  {id: 'stop', tableName: 'stops'},
+  {id: 'fare', tableName: 'fares'},
+  {id: 'feedinfo', tableName: 'feed_info'},
+  // FIXME: table name for calendar, fare, and schedule exception
+  {id: 'calendar', tableName: 'calendar'},
+  {id: 'scheduleexception', tableName: 'scheduleexception'},
+  {id: 'agency', tableName: 'agency'}
 ]
+
+export function getTableById (tableData: any, id: string, emptyArrayOnNull: boolean = true): any {
```
Comment (evansiroky) [inline lib/editor/util/gtfs.js]: I'm not quite sure why this is needed.  What is the reasoning behind this?

File: lib/editor/util/gtfs.js [hunk 5]
```diff
@@ -1,37 +1,88 @@
...
+//   const subEntityId = +routeParams.subEntityId
+//   return {
+//     ...routeParams,
+//     activeEntityId,
+//     subEntityId
+//   }
+// }
+
+// FIXME: Add flow.  Consolidate with gtfs/util/getEntityGraphQLRoot
+export const COMPONENT_LIST = [
```
Comment (evansiroky) [inline lib/editor/util/gtfs.js:40]: Use an object for this if it is needed.  There'd be no need to do array searches.

File: lib/editor/components/EditorInput.js [hunk 6]
```diff
@@ -264,26 +264,35 @@ export default class EditorInput extends Component {
...
-        const stopIndex = tableData.stop && tableData.stop.findIndex(s => s.id === activeEntity.id)
-        const stop = tableData.stop.find(s => s.id === currentValue)
-        const stops = [...tableData.stop]
-        // remove current entity from list of stops
-        // (because this is only used for parent_station selection and we can't make the self the parent)
+        // Create copy of stops table (because of destructive splice operation
+        // below).
+        const stops = [...getTableById(tableData, 'stop')]
+        const stopIndex = stops.findIndex(s => s.id === activeEntity.id)
+        const selectedStop = stops.find(s => s.id === currentValue)
```
Comment (evansiroky) [inline lib/editor/components/EditorInput.js]: I'd recommend creating an index of stopsByStopId to speed this up a bit.

File: lib/editor/components/map/AddableStopsLayer.js [hunk 7]
```diff
@@ -23,13 +31,15 @@ export default class AddableStopsLayer extends Component {
...
+      editSettings.addStops && zoom > 14 && bounds
     return (
       <FeatureGroup id='addable-stops-layer'>
-        {stops && activePattern && editSettings.addStops && zoom > 14 && bounds
+        {showAddableStops
           ? stops
             .filter(stop => {
               // filter out stops by lat/lng
-              if (stop.stop_lat > bounds.getNorth() || stop.stop_lat < bounds.getSouth() || stop.stop_lon > bounds.getEast() || stop.stop_lon < bounds.getWest()) {
+              if (this.isStopOutOfBounds(stop, bounds)) {
```
Comment (evansiroky) [inline lib/editor/components/map/AddableStopsLayer.js:42]: might be a good idea to put this in a selector

File: lib/editor/components/map/ControlPointsLayer.js [hunk 8]
```diff
@@ -1,69 +1,75 @@
...
+            }
+            const {point} = cp
+            const isActive = index === patternSegment
+            const color = isActive
+              ? '#000'
+              : index === patternSegment - 1 // control point before
+              ? 'blue'
+              : index === patternSegment + 1 // control point after
+              ? 'yellow'
+              : '#888' // inactive
```
Comment (evansiroky) [inline lib/editor/components/map/ControlPointsLayer.js:51]: Good idea to make this visual distinction.

File: lib/editor/components/map/DirectionIconsLayer.js [hunk 9]
```diff
@@ -4,43 +4,67 @@ import { divIcon } from 'leaflet'
...
+        : 10000
+      for (var i = 0; i < Math.floor(patternLength / iconInterval); i++) {
+        const distance = i ? iconInterval * i : iconInterval / 2
+        const position = along(patternLine, distance, {units: 'meters'})
+        if (!bounds) continue
+        if (
+          position.geometry.coordinates[1] > bounds.getNorth() ||
+          position.geometry.coordinates[1] < bounds.getSouth() ||
+          position.geometry.coordinates[0] > bounds.getEast() ||
+          position.geometry.coordinates[0] < bounds.getWest()
```
Comment (evansiroky) [inline lib/editor/components/map/DirectionIconsLayer.js:54]: Might be good to make a util function for whether a point is within a bounds object.

Comment (evansiroky) [reply lib/editor/components/map/DirectionIconsLayer.js:54]: Better yet, use [bounds.contains](http://leafletjs.com/reference-1.3.0.html#bounds-contains)

File: lib/editor/components/map/PatternsLayer.js [hunk 10]
```diff
@@ -62,25 +67,143 @@ export default class PatternsLayer extends Component {
...
+              }
             })
           : null
         }
       </FeatureGroup>
     )
   }
 }
+
+class PatternSegment extends Component {
```
Comment (evansiroky) [inline lib/editor/components/map/PatternsLayer.js:118]: Consider extracting this class into a new file.

File: lib/editor/containers/ActiveTripPatternList.js [hunk 11]
```diff
@@ -13,67 +13,80 @@ import {
...
+  // Trip pattern specific actions
+  addStopToPattern,
+  removeStopFromPattern,
+  resnapStops,
+  setActivePatternSegment,
+  setActiveStop,
+  setErrorMessage,
+  togglePatternEditing,
+  undoActiveTripPatternEdits,
+  updatePatternGeometry
```
Comment (evansiroky) [inline lib/editor/containers/ActiveTripPatternList.js:92]: 🥇

File: lib/editor/util/map.js [hunk 12]
```diff
@@ -226,230 +240,356 @@ export async function recalculateShape ({
...
+    const segment = patternSegments[i]
+    if (segment) {
+      // Add segment to list of shape points (last iteration should have no segment
+      // because the geometry should be capped with an extra control point).
+
+      // FIXME: this may not work for large arrays (100,000 or more)
+      shapePoints.push(...coordinatesToShapePoints(segment, lastDistance, shapePoints.length))
     }
   }
+  const resquencedShapePoints = shapePoints.map(resequenceShapePoints)
```
Comment (evansiroky) [inline lib/editor/util/map.js]: `resquencedShapePoints` is a typo

**Comment by landonreed** [comment]
After extensive testing, I am merging this branch into dev.

**Pull request merged**